src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageLoader.cs

```diff
@@ -60,12 +60,55 @@ public CompiledPageActionDescriptor Load(PageActionDescriptor actionDescriptor)
                 _applicationModelProviders[i].OnProvidersExecuted(context);
             }
 
-            for (var i = 0; i < _conventions.Length; i++)
+            ApplyConventions(_conventions, context.PageApplicationModel);
+
+            return CompiledPageActionDescriptorBuilder.Build(context.PageApplicationModel, _globalFilters);
+        }
+
+        internal static void ApplyConventions(
+            IPageApplicationModelConvention[] globalApplicationModelConventions,
+            PageApplicationModel pageApplicationModel)
+        {
+            for (var i = 0; i < globalApplicationModelConventions.Length; i++)
             {
-                _conventions[i].Apply(context.PageApplicationModel);
+                globalApplicationModelConventions[i].Apply(pageApplicationModel);
             }
 
-            return CompiledPageActionDescriptorBuilder.Build(context.PageApplicationModel, _globalFilters);
+            var applicationModelConventions = pageApplicationModel.HandlerTypeAttributes.OfType<IPageApplicationModelConvention>();
+            foreach (var convention in applicationModelConventions)
+            {
+                convention.Apply(pageApplicationModel);
+            }
+
+            var handlers = pageApplicationModel.HandlerMethods.ToArray();
+            foreach (var handlerModel in handlers)
+            {
+                var handlerModelConventions = handlerModel.Attributes.OfType<IPageHandlerModelConvention>();
+                foreach (var convention in handlerModelConventions)
```

Comment:
This needs to support global conventions too so they can be registered in startup.

Reply:
Sure, I can make this work for Pages. That said, we don't have an equivalent experience for controllers \ actions: https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/MvcOptions.cs#L77 only allows registering `IApplicationModelConvention`s not `Controller` \ `ActionModel` conventions. Ideally, we would have made `MvcOptions.Conventions` be a collection of a marker interfaces and then filter and apply accordingly. Do you think these are worthwhile issues to be solved for 3.0?

Reply:
Huh, I completely missed out on this - https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.Core/DependencyInjection/ApplicationModelConventionExtensions.cs#L170. Ignore my previous comment.